File: leetcode/golang/543_diameter_of_binary_tree.go
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func diameterOfBinaryTree(root *TreeNode) int {
	res := 0
	var dfs func(root *TreeNode, res *int) int
	dfs = func(root *TreeNode, res *int) int {
		if root == nil {
			return 0
		}
		left := dfs(root.Left, res)
		right := dfs(root.Right, res)
		*res = max(*res, left+right)
		return max(left, right) + 1
	}
	dfs(root, &res)
	return res
}

File: leetcode/c++/543_diameter_of_binary_tree.cpp
// https://leetcode.cn/problems/diameter-of-binary-tree

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int lenOfBinaryTree(TreeNode* root, int& res) {
        if (root == nullptr)
            return 0;
        int left = lenOfBinaryTree(root->left, res);
        int right = lenOfBinaryTree(root->right, res); 
        res = max(res, left+right);
        return max(left, right) + 1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int res = 0;
        lenOfBinaryTree(root, res);
        return res;
    }
};

// 只需要注意这条路径可能经过也可能不经过根节点 root 即可ac，然后就是记忆化搜索